src/async/pitfalls/blocking-executor.md

```diff
@@ -8,53 +8,41 @@ An easy workaround is to use async equivalent methods where possible.
-    join_all(std_sleep_futures).await;
-    assert!(now.elapsed().as_millis() >= 5050);
-
-    let now = Instant::now();
-    join_all(tokio_sleep_futures).await;
-    let runtime = now.elapsed().as_millis();
-    assert!((100..150).contains(&runtime));
+    let start = Instant::now();
+    let sleep_futures = (1..=10).map(|t| sleep_ms(&start, t, t * 10));
+    join_all(sleep_futures).await;
 }
-
 ```
 
 <details>
 
-- Using `std::thread::sleep` blocks the thread, so it prevents the executor from running. It means that while all futures are spawned at the same time, they all run one after the other. The runtime is the sum of all the `sleep` times. Try changing the runtime to `multi_thread` in a multi core environment to see how it impacts the run time.
-- A simple fix is to use `tokio::time::sleep`. Now, the `sleep` calls are `async` and they are properly scheduled by the executor.
-- Another fix would be to `tokio::task::spawn_blocking` which spawns an actual thread and transforms its handle into a future without blocking the executor. This thread is also scheduled as part of the executor's threadpool to grant better performance.
+* Run the code and see that the sleeps happen consecutively rather than
+  concurrently.
+
+* The `"current_thread"` flavor puts all tasks on a single thread. This makes the
+  effect more obvious, but the bug is still present in the multi-threaded
+  flavor.
+
+* Switch the `std::thread::sleep` to `tokio::time::sleep` and await its result.
+
+* Another fix would be to `tokio::task::spawn_blocking` which spawns an actual
+  thread and transforms its handle into a future without blocking the executor.
+  This is useful when calling external libraries that block the thread.
 
-- You should not think of tasks as OS threads. They do not map 1 to 1 and most executors will allow many tasks to run on a single OS thread. This creates multiple gotchas:
-  - For instance, using `std::sync::mutex` in an `async` runtime is very dangerous. When you lock the mutex in a thread then yield the executor using `.await` the thread might try to lock the mutex once more in a different task. Hence, prefer `async` alternatives like `tokio::sync::mutex`.
-  - Thread-local storage should also be used with care in async contexts as it doesn't map to specific tasks.
-  - Device drivers sometimes map to specific OS threads (for instance CUDA.) Prefer `tokio::task::spawn_blocking` when dealing with those.
-  - Some C libraries rely on thread local storage as well.
+* You should not think of tasks as OS threads. They do not map 1 to 1 and most executors will allow many tasks to run on a single OS thread. This creates multiple gotchas:
+  * For instance, using `std::sync::mutex` in an `async` runtime is very dangerous. When you lock the mutex in a thread then yield the executor using `.await` the thread might try to lock the mutex once more in a different task. Hence, prefer `async` alternatives like `tokio::sync::mutex`.
```

Comment:
I think this particular topic is debatable. A more practical advice is to use `std::sync::mutex` for short critical sections, as async mutexes have performance overhead.

Reply:
If you lock a sync Mutex and then use `.await` while the Mutex is locked, you're almost guaranteed to deadlock

Reply:
Yes. By a short critical section I meant one that does not include any `.await` or heavy computation.

Reply:
Yes, it's true. I'm maybe projecting my own bias but I would prefer people to default to async mutexes and use std Mutex only if there is a strong reason

Reply:
This is probably too detailed a point to get into in the course. I think understanding _why_ it might be a bad idea to use sync primitives in an async context is good enough. A good engineer can then figure out whether it makes sense to use `std::sync::mutex` in some specific situation. I'll try to scope this to explanation without making a recommendation.

Reply:
I agree. For me "very dangerous" sounded too strong. Perhaps just add that one should also consider the overhead of `tokio::sync::mutex`.